adrs/012-execution-domain-model.md

```diff
@@ -0,0 +1,138 @@
+# ADR 012 - Execution Domain Model
+### Status: Draft
+### Josh Holbrook
+
+## Context
+
+Matanuska contains a number of domain concepts to describe the syntax and
+semantics of source code, parsed trees, and compiled programs. Currently, they
+are named (and defined) as such:
+
+- **Expressions**: Units of code which apply a stack of *operations* on
+  *values* and evaluate to another value. Expressions do not cause side effects
+  directly, though function calls may execute *commands* which *do* cause
+  side effects.
+- **Values**: Elements within memory that represent numbers, booleans, strings,
+  and various object types. May either be contained within a variable, or
+  specified as a "literal" within source code.
+- **Operations** and **Operators**: Elements within *expressions* that add or
+  remove *values* from the expression stack when they're evaluated. These have
+  varying semantics, such as having infix/prefix/postfix operators and having
+  operator precedence. Note that *operators* are the syntactic element, and
+  *operations* are the corresponding units of execution.
+- **Commands**: Units of code which generally invoke state when executed.
+  Commands *may* take expressions as arguments, but may also include
+  non-expression syntactic elements or other commands. These
+  are generally separated by colons (`:`) within *command groups*.
+- **Interactive commands**: Commands which may only be executed in an
+  interactive session, through use of the `command` module. These commands
+  are syntactically within non-line *command groups*.
+- **Runtime commands**: Commands which are compiled into chunks and op codes
+  and executed by the `runtime`.
+- **Simple** and **complex** commands: Complex commands are commands
+  which *may* contain other commands. This is in contrast to "simple commands",
+  which *never* contain other commands. This concept is not currently
+  represented within the source code, but is useful to distinguish in the
+  context of this ADR.
+- **Command groups**: A series of commands, separated by colons (`:`). These
+  command groups may form the non-numbered portion of a line, or may constitute
+  "bare" *interactive commands*.
+- **Lines**: A command group combined with a preceding line number. Lines are
+  currently contained on a single source line, though that may not be true in
+  the future if newlines can be escaped (like Bash).
+
+**TODO:** Instructions are actually made up of op codes *and* arguments. These
```

Comment:
Arguably, bytecode, the AST and source code are all different "languages", and can have distinct concepts of "instructions" and "operations". This might not be the problem I think it is.